src/source-dpdk.c

```diff
@@ -0,0 +1,781 @@
+    uint64_t usec = CyclesToMicroseconds(cycles) + orig_tv->tv_usec;
+    new_tv->tv_sec = orig_tv->tv_sec + usec / 1000000;
+    new_tv->tv_usec = (usec % 1000000);
+}
+
+static void DPDKSetTimevalOfMachineStart(struct timeval *tv)
+{
+    gettimeofday(tv, NULL);
+    tv->tv_sec -= DPDKGetSeconds();
+}
+
+/**
+ * Initializes real_tv to the correct real time. Adds TSC counter value to the timeval of
+ * the machine start
+ * @param machine_start_tv - timestamp when the machine was started
+ * @param real_tv
+ */
+static void DPDKSetTimevalReal(struct timeval *machine_start_tv, struct timeval *real_tv)
+{
+    CyclesAddToTimeval(rte_get_tsc_cycles(), machine_start_tv, real_tv);
+}
+
+/* get number of seconds from the reset of TSC counter (typically from the machine start) */
+static uint64_t DPDKGetSeconds()
+{
+    return CyclesToSeconds(rte_get_tsc_cycles());
+}
+
+/**
+ * Attempts to retrieve NUMA node id on which the caller runs
+ * @return NUMA id on success, -1 otherwise
+ */
+static int GetNumaNode(void)
+{
+    int cpu = 0;
+    int node = -1;
+
+#if defined(__linux__)
+    cpu = sched_getcpu();
+    node = numa_node_of_cpu(cpu);
+#else
+    SCLogWarning(SC_ERR_TM_THREADS_ERROR, "NUMA node retrieval is not supported on this OS.");
+#endif
+
+    return node;
+}
+
+static int i40eDeviceSetRSSQueues(struct rte_eth_rss_conf rss_conf, int port_id, int nb_rx_queues)
+{
+    struct rte_flow_action_rss rss_action_conf = { 0 };
+    struct rte_flow_attr attr = { 0 };
+    struct rte_flow_item pattern[] = { { 0 }, { 0 }, { 0 }, { 0 } };
+    struct rte_flow_action action[] = { { 0 }, { 0 } };
+    struct rte_flow *flow;
+    struct rte_flow_error flow_error = { 0 };
+    uint16_t queues[RTE_MAX_QUEUES_PER_PORT];
+
+    for (int i = 0; i < nb_rx_queues; ++i)
+        queues[i] = i;
+
+    rss_action_conf.func = RTE_ETH_HASH_FUNCTION_DEFAULT;
+    rss_action_conf.level = 0;
+    rss_action_conf.types = 0; // queues region can not be configured with types
+    rss_action_conf.key = rss_conf.rss_key;
+    rss_action_conf.key_len = rss_conf.rss_key_len;
+    rss_action_conf.queue_num = nb_rx_queues;
+    rss_action_conf.queue = queues;
+
+    attr.ingress = 1;
+    pattern[0].type = RTE_FLOW_ITEM_TYPE_END;
+    action[0].type = RTE_FLOW_ACTION_TYPE_RSS;
+    action[0].conf = &rss_action_conf;
+    action[1].type = RTE_FLOW_ACTION_TYPE_END;
+
+    flow = rte_flow_create(port_id, &attr, pattern, action, &flow_error);
+    if (flow == NULL) {
+        SCLogError(SC_ERR_DPDK_CONF, "Create errror: %s", flow_error.message);
+        int ret = rte_flow_validate(port_id, &attr, pattern, action, &flow_error);
+        SCLogError(SC_ERR_DPDK_CONF, "Err on flow validation: %s \n errmsg: %s", rte_strerror(-ret),
+                flow_error.message);
+        return ret;
+    } else {
+        SCLogInfo("RTE_FLOW queue region created");
+    }
+    return 0;
+}
+
+static int i40eDeviceCreateRSSFlow(int port_id, struct rte_eth_rss_conf rss_conf, uint64_t rss_type,
```

Comment:
can this be replaces by enabling the i40e symmetric hash support?

http://patches.dpdk.org/project/dpdk/patch/1406184149-11531-5-git-send-email-helin.zhang@intel.com/

Reply:
I was unable to verify your patch but considering it's from 2014, it'll most likely be deprecated. Similarly, I have found a solution how to set symmetric RSS on i40e on DPDK versions below 20.xx.
https://haryachyy.wordpress.com/2019/01/18/learning-dpdk-symmetric-rss/

This can be used but is already flagged deprecated in 19.xx. I've decided to use it anyway for the version 19.xx (the least version targetted by the PR) and for DPDK 20.xx+ I use the rte_flow. According to my knowledge, it's the current preffered way how to set RSS on i40e driver. 

I40e has pretty peculiar behavior when I'm trying to set RSS. It took me a decent amount of time to figure out how to set RSS via rte_flow so I made a little example out of it.
https://github.com/lukashino/i40e-symmetric-rss-rte-flow